psignifit/getConfRegion.py

```diff
@@ -48,12 +48,111 @@
 
 from .marginalize import marginalize
 
 
+def confidence_intervals(probability_mass: np.ndarray, grid_values: np.ndarray,
+                         p_values: Sequence[float], mode: str) -> np.ndarray:
+    """ Confidence intervals on probability grid.
+
+    Supports two methods:
+
+        - 'project', projects the confidence region down each axis.
+          Implemented using :func:`grid_hdi`.
+        - 'percentiles', finds alpha/2 and 1-alpha/2 percentiles (alpha = 1-p_value).
+          Implemented using :func:`percentile_intervals`.
+
+    Args:
+        probability_mass: Probability mass at each grid point, shape (n_points, n_points, ...)
+        grid_values: Parameter values along grid axis in the same order as probability_mass dimensions,
+                     shape (n_dims, n_points)
+        p_values: Probabilities of confidence in the intervals.
+        mode: Either 'project' or 'percentiles'.
+    Returns:
+        Start and end grid-values for the confidence interval
+        per dimension and p_value, shape (n_dims, n_p_values, 2)
+    Raises:
+        ValueError for unsupported mode or sum(probability_mass) != 1.
+     """
+    CI_METHODS = { 'project': grid_hdi, 'percentiles': percentile_intervals }
+    if mode in CI_METHODS:
+        calc_ci = CI_METHODS[mode]
+    elif mode == 'stripes':
+        raise ValueError('Confidence mode "stripes" is not supported anymore. "' 
+                         f'"Use one of {list(CI_METHODS.keys())}.')
+    else:
+        raise ValueError(f'Expects mode as one of {list(CI_METHODS.keys())}, got {mode}')
+
+    if not np.isclose(probability_mass.sum(), 1):
+        raise ValueError(f'Expects sum(probability_mass) to be 1., got {probability_mass.sum():.4f}')
+    intervals = np.empty((probability_mass.ndim, len(p_values), 2))
+    for p_ix, p_value in enumerate(p_values):
+        intervals[:, p_ix, :] = calc_ci(probability_mass, grid_values, p_value)
+
+    return intervals
+
+
+def grid_hdi(probability_mass: np.ndarray, grid_values: np.ndarray, credible_mass: float) -> np.ndarray:
+    """ Highest density intervals (hdi) on a grid.
+
+    Calculates the grid region, in which the probability mass
+    is at least the credible mass (highest density region, hdr).
+    This region is projected to the individual dimensions.
+    The intervals range from the first to the last grid entry
+    in the projected hdr.
+
+    See `stats.stackexchange` for an explanation of the method.
+
+    Args:
+        probability_mass: Probability mass at each grid point, shape (n_points, n_points, ...)
+        grid_values: Parameter values along grid axis, shape (n_dims, n_points)
+        credible_mass: Minimal mass in highest density region
+    Returns:
+        Grid value at interval per dimension, shape (n_dims, 2)
+
+    .. _stats.stackexchange: https://stats.stackexchange.com/questions/148439/what-is-a-highest-density-region-hdr
+    """
+    decreasing_mass = np.sort(np.ravel(probability_mass))[::-1]
```

Comment:
Use `a.reshape(-1)` instead of `a.ravel()` everytime we need a view (`ravel` copies the underlying data). Look at the `np.ravel` documentation for more info:

> When a view is desired in as many cases as possible, `arr.reshape(-1)` may be preferable.